### ISSUE TITLE
Split DNS does not respect local routing table on macOS

### ISSUE THREAD

**Issue by torarnv** [tailscale/tailscale#12940]
### What is the issue?

I've set up split DNS for `my.domain` to point to a LAN IP `192.168.1.2`, where I run a `coredns` DNS server on a Synology NAS as a docker container with `--net=host`.

Resolving `foo.my.domain` from my laptop, where I've enabled "Use Tailscale DNS settings" correctly ends up querying the `coredns` server:

```
coredns  | [INFO] 192.168.1.50:61304 - 34598 "A IN foo.my.domain. udp 42 false 4095" NOERROR qr,rd,ra,ad 127 0.047305328s
```
And I can see the source IP for the DNS request matching my laptop (`192.168.1.50`).

Next, I add a subnet route to my tailnet so that I can access the DNS server when I'm outside my LAN. Following the [documentation](https://tailscale.com/kb/1023/troubleshooting#how-to-prioritize-lan-traffic-with-overlapping-subnet-routes) I add a `192.168.0.0/23` subnet route, so that it will be prioritized below the direct route (`192.168.0.0/24`) when I'm on my LAN. I don't want all traffic from my laptop to my NAS to go through Tailscale when I'm on my LAN, as that hurts performance and throughput. 

`❯ sudo tailscale set --advertise-routes "192.168.0.0/23" --snat-subnet-routes=false`

I disable SNAT following the [documentation](https://tailscale.com/kb/1019/subnets#disable-snat), so that I can see the original IP when routed through the subnet router.

The down-prioritizaion of the subnet route seems to work as expected from my laptop:

```
❯ route -n get 192.168.1.2
   route to: 192.168.1.2
destination: 192.168.1.2
  interface: en0
      flags: <UP,HOST,DONE,LLINFO,WASCLONED,IFSCOPE,IFREF>
 recvpipe  sendpipe  ssthresh  rtt,msec    rttvar  hopcount      mtu     expire
       0         0       180         0        22         0      1500      1199
```

However, if I try to resolve `bar.my.domain` in this setup, the Tailscale DNS server on `100.100.100.100` seems to ignore the local route priority and routes the request through the tailnet, resulting in `coredns` seeing:

```
coredns  | [INFO] 192.168.1.2:63228 - 61478 "A IN bar.my.domain. udp 42 false 4095" NOERROR qr,rd,ra,ad 127 0.005832224s
```

Note how the source IP is now the IP of the machine the DNS server is running on, and where Tailscale is advertising the subnet route, instead of the machine making the DNS request. The disabled SNAT doesn't seem to have any effect on this for some reason.

In other words, I can no longer use the source IP to determine where the request is coming from, for example to enable split horizon DNS based on whether I'm on my LAN or not (as a workaround for #3506).

If I disable "Use Tailscale subnets" on my laptop, things work as expected, but I don't want to have to toggle this setting on and off when moving in and out of my LAN. Plus, I wouldn't be able to toggle it off if I ever add subnet routes to other networks outside of my LAN.

I have also tied to add the Tailscale IP of the machine running the DNS server to Tailscale's Split DNS settings instead of the LAN address, but that results in `coredns` always reporting the source IP as `127.0.0.1`.

### Steps to reproduce

1. Add split DNS, pointing to LAN address
2. Add subnet route that covers that address, but at a lower priority than the LAN net mask
3. Resolve hostname via the custom DNS
4. Observe lack of accurate source IP on DNS server

### OS

macOS, Synology

### OS version

Synology DSM 7, macOS 14.5

### Tailscale version

1.70.0

### Bug report

BUG-63601a056145aba07eae2b269b666e8c5bbea5309f98e682d8788e89ca774659-20240728182343Z-3332f938dbd7089a

**Comment by torarnv** [comment]
I tried to reproduce what I was seeing without DNS in play, so I ran Caddy on the Synology machine:

`❯ docker run -it --rm --net=host caddy caddy file-server -b -a -l :8080`

Accessing `http://192.168.1.2:8080` from my laptop, the access log show:

```
2024/07/28 18:38:00.439	INFO	http.log.access	handled request	{"request": {"remote_ip": "192.168.1.24", "remote_port": "57721", "client_ip": "192.168.1.24", "proto": "HTTP/1.1", "method": "GET", "host": "192.168.1.2:8080", "uri": "/" ...
```

But if I access the same URL from my iPhone on 5G with Tailscale connected I see:

```
2024/07/28 18:39:18.218	INFO	http.log.access	handled request	{"request": {"remote_ip": "192.168.1.2", "remote_port": "57166", "client_ip": "192.168.1.2", "proto": "HTTP/1.1", "method": "GET", "host": "192.168.1.2:8080", "uri": "/",  ...
```

So there seems to be two issues at play here:

1. The Split DNS forwarding in the Tailscale DNS server on `100.100.100.100` does not respect the route priority of the host
2. When the routing then happens via the Tailscale subnet route `192.168.0.0/23`, the disabling of SNAT does not seem to work, so the original IP is lost.

**Comment by torarnv** [comment]
> 2. When the routing then happens via the Tailscale subnet route `192.168.0.0/23`, the disabling of SNAT does not seem to work, so the original IP is lost.

This seems to be an issue for other people as well, based on comments in #1995

**Comment by gruselhaus** [comment]
Could the Tailscale team share an update on this issue and any planned fix or supported workaround?

This is a blocker for our office/remote-access design in a larger enterprise environment on the **Tailscale Enterprise plan**. We use AD-integrated DNS, and the AD namespace also contains internal application names that must resolve both on the office LAN and over Tailscale.

Our intended setup keeps Tailscale connected and subnet routes enabled. In the office, more-specific LAN routes should take precedence over the advertised Tailscale subnet route, including for requests to the internal DNS server. Outside the office, the same DNS server should be reachable through the subnet router. We need to avoid making otherwise locally reachable office DNS dependent on the Tailscale subnet router.

The macOS client version we assessed is **1.102.4**. Reviewing [`dnsIPsOverTailscale` and `fwdDNSLinkSelector.PickLink` in that version](https://github.com/tailscale/tailscale/blob/v1.102.4/wgengine/userspace.go#L1498-L1550) suggests the forwarding path can still explicitly select the tunnel interface when the resolver is covered by a Tailscale route. A packet-capture reproduction with our proposed more-specific LAN routes on this version is still pending.

Could you clarify:

1. Is the behavior reported here still considered a bug, and is a fix planned or available in a particular release?
2. Is there a supported way for Split DNS forwarding on macOS to honor the preferred LAN route while Tailscale remains connected and accepts subnet routes?
3. What workaround do you recommend for Enterprise deployments that need automatic office/off-site transitions, without manually toggling subnet acceptance or introducing additional DNS proxies?

Thank you. Guidance on the supported design would help us unblock this deployment.
